### ISSUE TITLE
Decide which SSR Data Fetching Library to use

### ISSUE THREAD

**Issue by Shub1427** [freeCodeCamp/chapter#258]
List of Libraries that has been discussed till date:

* isomorphic-fetch
* isomorphic-unfetch
* axios
* cross-fetch
* node-fetch
* whatwg-fetch

> Last two libraries, if used, should be used together, as they are not isomorphic in nature.

@Zeko369 suggested, respective to PR https://github.com/freeCodeCamp/chapter/pull/140, we are already using `isomorphic-fetch`

@Sonicrida suggested `axios` is not maintained for a while and has some security vulnerabilities.
@Shub1427  concluded that `isomorphic-fetch`, `isomorphic-unfetch` and `axios` hasn't been maintained for a long period, though recently from September `axios` has a new core developer helping to resolve issues as fast as he could.

Regarding security vulnerability for axios, it was a false alarm as per [this](https://github.com/axios/axios/issues/2447).

@nik-john suggested to use `cross-fetch` which seems a viable option, as it's being actively maintained, and is a wrapper around `node-fetch` and `whatwg-fetch` (which I was trying to do by myself in PR https://github.com/freeCodeCamp/chapter/pull/166).

_This discussion needs to be concluded, as the PR https://github.com/freeCodeCamp/chapter/pull/166 is been on hold for a long time now due to this and acts as a blocker to the same._

**Comment by allella** [comment]
Pinging @vaibhavsingh97 @joelrozen @ThomasRoest @xarielx @Haxified for help resolving this decision.

**Comment by nik-john** [comment]
![Screenshot_20191124-100832_Chrome](https://user-images.githubusercontent.com/1117182/69497416-6af2a880-0ea2-11ea-96c1-7019bd7a8b05.jpg)

I personally use `node-fetch` on my projects but my arguments for `cross-fetch` are that it seems to be well maintained, but it looks like it's not as widely adopted yet.

**Comment by joelrozen** [comment]
Another one to consider is SWR (https://github.com/zeit/swr) which, given we are using zeit's next.js anyway, may be the better way to go.

**Comment by Zeko369** [comment]
@joelrozen SWR is a wrapper around any fetch function so that isn't what we're looking for, since we still need to provide a `fetcher` function
![Screenshot_20191124-232832_Chrome](https://user-images.githubusercontent.com/3064377/69502696-68c83280-0f12-11ea-8ae5-c9289489155d.jpg)

**Comment by joelrozen** [comment]
@Zeko369 ahhhh ooops. 
Okay, well I'll go with whatever every one else decides, since I've only had experience with axios anyway and that seems to be out, so no opinion either way on the other choices.

**Comment by Shub1427** [comment]
Personally, I would like to work with `node-fetch` and `whatwg-fetch`. They are well maintained and widely used as well.

`cross-fetch` is a wrapper around above two libs, which we can do ourselves or use `cross-fetch` instead.

NextJS supports conditional imports:
* https://github.com/zeit/next.js/issues/2473?#issuecomment-362119102
* Or we can also do something like https://github.com/zeit/next.js/issues/7755#issuecomment-508633125, where webpack configs is changed, but I don't think this will work for our case, as nextjs/webpack won't recompile the app while rendering it on Server and sending it as a serialized string response to browser.

**Comment by Zeko369** [comment]
+1 `cross-fetch`

**Comment by allella** [comment]
There seems to be support for from @Zeko369, @Shub1427 @nik-john about cross-fetch (a wrapper around node-fetch and whatwg-fetch)

@Sonicrida @joelrozen any opposition?

**Comment by Haxified** [comment]
cross-fetch 100%

**Comment by Zeko369** [comment]
@timmyichen

**Comment by timmyichen** [comment]
I'm pretty indifferent about which library is chosen as long as it fulfills all our needs

**Comment by Vitao18** [comment]
I'd prefer to use cross-fetch

**Comment by nik-john** [comment]
Ok closing this as `cross-fetch` seems to be the consensus

@Shub1427 Hopefully this unblocks you

**Comment by bernhard-hofmann** [comment]
Do we have the criteria for selection?

There seem to be a lot of opinions and they're based on a variety of criteria. If we start with a prioritised list of requirements and desires I think we'll find it easier to reach an objective consensus.

**Comment by nik-john** [comment]
@bernhard-hofmann It's pretty straight forward - because we're going to use this library extensively on Chapter (in no particular order):

- It needs to be a robust library with no glaring open issues
- It needs to perform all functions of an HTTP library
- It has to be actively maintained
- It needs to work seamlessly on the server as well as the client
- It should have a hassle-free integration with NextJS
